### ISSUE TITLE
Improve severly lacking Trainable.fit() documentation

### ISSUE THREAD

**Issue by hnsl** [tensorflow/tensorflow#6067]
I'm a beginner, some questions that came to mind after reading it multiple times:

- Give examples of x and y, I never got those to work. It either breaks because they have no "dtype" or complains that I should use input_fn instead.
- Are x/y supposed to be tensors?
- When is input_fn called?
- If input_fn returns some kind of "queue" how is it used?
- What would you actually expect a "step" to do? (Needed to know estimate a good value of "steps")
- How many training examples does each step consume/look at?
- Why can't batch_size be specified together with input_fn?
- What does batch_size actually do? What is a batch?
- Do you always want to do training with limited batch size?

Thanks.

**Comment by poxvoculi** [comment]
Sorry you're having a hard time.  I've been through that too.

This forum is for bug reports, not use questions, which are better addressed on stackoverflow.

If the documentation is poor, that counts as a bug, but in this case the python documentation for learn.Trainable.fit is on a consistent level of detail with the documentation throughout.  It sounds like you need a gentle introduction to TensorFlow, because your questions are largely about basic concepts.  I would suggest working through some of the basic tutorials first, like maybe https://www.tensorflow.org/versions/r0.10/tutorials/mnist/tf/index.html

A confusing aspect of TensorFlow is that one typically defines programs in python, but most of the real computation happens in a C++ -built backend runtime.   The python program execution builds up an Op graph, then passes it to the backend along with input data, requests execution, and waits for the result.  Usually a 'step' is a single iteration of running the TensorFlow Op graph for some inputs and updating any parameters that are being trained.  In the backend all input/output values are Tensors, but in the python program they are python data types, which sometimes but not always correspond to Tensors in the backend.  After working through some tutorials, hopefully this will start to make more sense.

**Comment by hnsl** [comment]
I've already been introduced to TF so I would say that I'm aware of the absolute basics and the things you mention here. You can consider this a bug report of the fit() documentation, rather than a problem I am specifically having. I'm already using fit() today, but mostly through trial and error. I've read most of the TF documentation without finding many examples or explanations that answers my questions, that could improve my use of fit().

I don't think fit() is on a "consistent level of detail with the documentation throughout", unless you  measure consistency in something trivial like "same amount of characters". Even in that case I think it should not just be that, it should be great in general.

In my opinion any API should carefully document:
- Guarantees
- Constraints
- Intended use
- Important design motivations and historical changes

Considering how important fit() actually is for users to easily implement ML, I can see that it easily need twice as much detail. I think most people would currently need plenty of context to just understand some of the phrases used in the documentation.

Here's some examples:

      x: Matrix of shape [n_samples, n_features...] or the dictionary of Matrices.
         Can be iterator that returns arrays of features or dictionary of arrays of features.
         The training input samples for fitting the model. If set, `input_fn` must be `None`.

What's n_samples and n_features? Why the ... and how is it obvious how that sequence would continue? What's "the dictionary of Matrices"? Where is that term defined? Why "arrays of features" and not just "an array of features"? Do you mean "iterator over arrays of features"? Can the iterator also return a dictionary or do you mean that x can also be a dictionary? Why is that in the same sentence? What is the key in the mentioned dictionaries, how are they indexed? When it says "The training input samples for fitting the model." what exactly is an "input sample" and what is it not? Is it a feature or a label? What actual datatypes/values is acceptable for an input matrix contains?

Throughout the documentation there is also no specified reason why both "x" and "input_fn" exists. Is x just a historical artifact? Should you always use an input_fn? What is the purpose of "x" if you can have an "input_fn"?

That's just the first argument, I could continue, but I think this makes my point clear. Note that I can probably make an educated guess and be right for some of these questions. But I think the documentation should be better than that. I think the documentation should be clear enough that usage doesn't require guesswork or dissecting examples.

I could accept the argument that the documentation is so general and context-full because the trainable class itself is so general that it's hard to be specific. But in this case classes like DNNRegressor need to have their own documentation on their fit() implementation and not just refer to the trainable documentation.

That said, I could have missed some topic that describe fit() better. I might also have unrealistic expectations. I apologize if that is the case.

**Comment by martinwicke** [comment]
Your expectations are not unrealistic, but we are talking about a component in contrib. Its interface is still evolving (you'll see that looking at head), and its documentation is not complete. You can use the examples in the tutorials section as a starting point, for example:
https://www.tensorflow.org/versions/r0.12/tutorials/tflearn/index.html#fit-dnnclassifier
https://www.tensorflow.org/versions/r0.12/tutorials/input_fn/index.html#building-input-functions-with-tf-contrib-learn

x and y are supposed to be numpy arrays, or dictionaries of numpy arrays. If you use an input_fn, the return values of said function should be dictionaries of tensors containing a single batch of data when evaluated.

I will close this bug, we will rework the documentation soon.
